### PR TITLE
Race changes starting ability scores and hide increment and decrement on max and min values

### LINKED ISSUE #115: Race selected at start alters default starting ability scores
**User Story**
As a game developer, I'd like to have the race the player initially selects to change their default staring ability scores, so that there is variable is game play and character builds.

**Acceptance Criteria**

- If the human race is selected, the starting ability scores for strength and intelligence should increase by 2
- If the elf race is selected, the starting ability scores for dexterity and charisma should increase by 2
- If the dwarf race is selected, the starting ability scores for constitution and wisdom should increase by 2

**Notes**

---

**Why is this feature needed? Please describe the problem your requested feature wants to solve**

The player can currently select a race when starting but it has no affect on game play currently.

**Describe the solution you'd like**

The race selected at the start affects their starting default ability score distribution.

### LINKED ISSUE #123: Hide increment and decrement ability scores when at minimum and maximum values
**User Story**
As a game developer, I'd like to hide the increment and decrement of an ability score if it is at either its minimum or maximum values, so that players are informed of when they reach the minimum of maximum value for the ability stat they are investing in.

**Acceptance Criteria**

- If an ability score it at 0, the decrement button is not displayed
- If an ability score it at 20, the increment button is not displayed

**Notes**

---

**Why is this feature needed? Please describe the problem your requested feature wants to solve**

To inform players of when they reach the minimum or maximum ability score for a stat.

**Describe the solution you'd like**

The decrement button is not displayed when its equivalent ability score is at 0 and the increment button is not displayed when its equivalent ability score is at 20.

### PR DESCRIPTION
**Pull request by Frosty273**
GitHub Issue (If applicable): #115 & #123 

## PR Type

What kind of change does this PR introduce?

## What is the current behavior?
1. The race a player selects during character creation has no affect on the game.
2. The arrow button to increment and decrement an ability score remains even if the button can no longer increase or decrease the ability score value.

## What is the new behavior?
1. The race a player selects during character creation alters the starting ability scores and minimum possible value for their respective abilities.
2. The arrow button to increment and decrement an ability score is hidden if the ability score value is currently at either its maximum or minimum possible value.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current upstream repo
- [ ] Added new tests to the test suite for the feature
- [ ] Tested the full test suite, and ensured that the feature does not break current build.
- [ ] Docs have been added/updated which fit (for bug fixes / features)
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal)
- [ ] Checked licensing of Frameworks/Libraries (if applicable) 

## Other information

Internal Issue (If applicable):

Closes #115 
Closes #123

### CODE REVIEW

File: src/features/dialog-manager/reducer.js [hunk 1]
```diff
@@ -40,14 +42,40 @@ const initialState = {
...
 const dialogManagerReducer = (state = initialState, { type, payload }) => {
     const { abilities, abilities_minimum, character } = state;
+    // if in character creation mode
+    if (state.characterCreation && character.characterName !== ' ') {
+        resetAbilityScoreValues(abilities);
+        switch (character.characterRace) {
+            case 'Human':
+                abilities.strength = RACE_BONUS_STARTING_ABILITY_SCORE_VALUE;
+                abilities.intelligence = RACE_BONUS_STARTING_ABILITY_SCORE_VALUE;
+                abilities_minimum.min_strength = 2;
```
Comment (KimberleyEvans-Parker) [inline src/features/dialog-manager/reducer.js]: It might be better to  have the `ABILITY_BONUS = 2` defined as a constant rather than the `RACE_BONUS_STARTING_ABILITY_SCORE_VALUE`.  

That way you could have `abilities_minimum.min_strength = ABILITY_BONUS;` and `abilities.strength = STARTING_ABILITY_SCORE_VALUE + ABILITY_BONUS;` (or `abilities.strength += ABILITY_BONUS;`)

**Review by KimberleyEvans-Parker — CHANGES_REQUESTED** [review]
Visually looks great! Just the code style change I mentioned and you're good.

File: src/features/dialog-manager/reducer.js (hunk 1 above)
Comment (Frosty273) [reply src/features/dialog-manager/reducer.js]: Will do, the minimum values are all manually set to 0 in the initial state, should we convert those to constants as well?

File: src/components/ability/index.js [hunk 2]
```diff
@@ -4,12 +4,15 @@ import Button from '../button';
...
-const Ability = ({ name, value, increment, decrement }) => {
+const Ability = ({ name, value, minValue, increment, decrement }) => {
     return (
         <>
             <div className="ability-score__container">
                 <span className="ability-score__text">{name}:</span>
-                <div className="ability-score__button ">
+                <div
+                    className="ability-score__button"
+                    style={{ visibility: value === 20 ? 'hidden' : 'visible' }}
```
Comment (Matteas-Eden) [inline src/components/ability/index.js]: For any future feature that may increase the maximum beyond 20, I think it would be good to have a `maxValue` here instead of a hardcoded value.

File: src/components/ability/index.js [hunk 3]
```diff
@@ -19,7 +22,12 @@ const Ability = ({ name, value, increment, decrement }) => {
...
-                <div className="ability-score__button ">
+                <div
+                    className="ability-score__button"
+                    style={{
+                        visibility: value === minValue ? 'hidden' : 'visible',
+                    }}
+                >
                     <Button
                         title=" "
                         icon="caret-left"
```
Comment (Matteas-Eden) [inline src/components/ability/index.js:37]: I know this isn't actually part of this PR, but I think this qualifies for a refactor going forward. Passing in a blank title is nonsensical. Probably not in this PR, but I just wanted to point it out.

File: src/features/dialog-manager/reducer.js [hunk 4]
```diff
@@ -40,14 +42,40 @@ const initialState = {
...
+            case 'Dwarf':
+                abilities.constitution += RACE_ABILITY_BONUS;
+                abilities.wisdom += RACE_ABILITY_BONUS;
+                abilities_minimum.min_constitution = RACE_ABILITY_BONUS;
+                abilities_minimum.min_wisdom = RACE_ABILITY_BONUS;
+                break;
+            default:
+        }
+    }
+
```
Comment (Matteas-Eden) [inline src/features/dialog-manager/reducer.js]: Could this be made into a function, or refactored in some way? It seems very out of place where it is right now. That being said, I'm not entirely sure what change would be required to improve it and make the code cleaner whilst preserving the functionality.

**Review by Matteas-Eden — CHANGES_REQUESTED** [review]
It looks good, just the one change about `maxValue` that I'd like addressed. The other comments are just my musings and don't need to be addressed for me to approve the PR.

File: src/components/ability/index.js (hunk 2 above)
Comment (Pyxxil) [reply src/components/ability/index.js]: The max value already exists as a constant (MAX_ABILITY_SCORE).

Comment (Matteas-Eden) [reply src/components/ability/index.js]: While that does exist, the implementation I'm suggesting would allow the maximum ability score value to be changed at runtime. Perhaps it should be part of a separate PR, in which case line 14 should be changed to use `MAX_ABILITY_SCORE`.

Comment (KimberleyEvans-Parker) [reply src/components/ability/index.js]: Why would the max ability score need to be changed at runtime? 
I think that the 20 should probably be changed to `MAX_ABILITY_SCORE`

**Review by KimberleyEvans-Parker — CHANGES_REQUESTED** [review]
Just the one change from me - line 14 in ability/index (sorry I didn't pick it up before).  May as well use the `const MAX_ABILITY_SCORE`

File: src/features/dialog-manager/reducer.js (hunk 1 above)
Comment (Pyxxil) [reply src/features/dialog-manager/reducer.js]: I think 0's should be fine (unless we ever plan on changing that value).

File: src/features/dialog-manager/reducer.js (hunk 4 above)
Comment (Frosty273) [reply src/features/dialog-manager/reducer.js]: What part of it is out of place for you? The switch case or the repeated 4 lines under a case? Depending on where the issue is, I can refactor into a function probably from there.

File: src/utils/reset-starting-abilities.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+    abilities.dexterity = STARTING_ABILITY_SCORE_VALUE;
+    abilities.charisma = STARTING_ABILITY_SCORE_VALUE;
+    abilities.intelligence = STARTING_ABILITY_SCORE_VALUE;
+    abilities.wisdom = STARTING_ABILITY_SCORE_VALUE;
+    abilities_minimum.min_strength = 0;
+    abilities_minimum.min_constituion = 0;
+    abilities_minimum.min_dexterity = 0;
+    abilities_minimum.min_charisma = 0;
+    abilities_minimum.min_intelligence = 0;
+    abilities_minimum.min_wisdom = 0;
```
Comment (Matteas-Eden) [inline src/utils/reset-starting-abilities.js:15]: Could this be refactored to make use of a loop? Or perhaps replace 0 with a macro? (I know it seems silly, I just want to reduce magic numbers)

File: src/features/dialog-manager/reducer.js [hunk 6]
```diff
@@ -40,14 +42,20 @@ const initialState = {
...
+        characterName: ' ',
         characterRace: 'Human',
         characterClass: 'Fighter',
     },
 };
 
 const dialogManagerReducer = (state = initialState, { type, payload }) => {
     const { abilities, abilities_minimum, character } = state;
+    // if in character creation mode
+    if (state.characterCreation && character.characterName !== ' ') {
```
Comment (Matteas-Eden) [inline src/features/dialog-manager/reducer.js]: Could change the second condition to `character.characterName.trim()`

File: src/features/dialog-manager/reducer.js [hunk 7]
```diff
@@ -40,14 +42,20 @@ const initialState = {
     },
     characterCreation: false,
     character: {
-        characterName: '',
+        characterName: ' ',
```
Comment (Matteas-Eden) [inline src/features/dialog-manager/reducer.js]: Is this change needed for some particular reason?

**Review by Matteas-Eden — DISMISSED** [review]
LGTM. I have a few comments about potential refactoring, but they're not important and are non-blocking changes.

**Review by Matteas-Eden — APPROVED** [review]
Thanks for addressing my feedback, LGTM

**Review by KimberleyEvans-Parker — APPROVED** [review]
Looks nice. Great work :)
